packages/@aws-cdk/aws-events/lib/rule.ts

```diff
@@ -127,14 +127,11 @@ export class Rule extends Resource implements IRule {
   public addTarget(target?: IRuleTarget) {
     if (!target) { return; }
 
-    const targetProps = target.bind(this);
-    const id = sanitizeId(targetProps.id);
-    const inputProps = targetProps.input && targetProps.input.bind(this);
+    // Simply increment id for each `addTarget` call. This is guaranteed to be unique.
+    const id = `Target${this.targets.length}`;
```

Comment:
I'm not sure this will work correctly if we update a list of 2 elements to a list of 2 different elements.

I believe what will happen is CloudFormation will call `addTarget("Rule1", { ... })` (the new Rule1), and then `removeTarget("Rule1")` (intending to remove the old rule1).

Anything that depends on addition order is super scary to me. I'd rather have a different consistent way of generating IDs.

Reply:
Either an optional "id" field on every target that can be used in lieu of the default id, or we hash in the target configuration to make the ID unique.

Reply:
Are you sure? CF will likely use [PutTargets](https://docs.aws.amazon.com/AmazonCloudWatchEvents/latest/APIReference/API_PutTargets.html) which updates everything in one go?

Reply:
OK I get what you mean, let's hash the configuration.

Reply:
How do you suggest passing the id to the `EcsTask` target (which needs the id for its `AwsCustomResource`) if we hash it in `aws-events` after receiving the full configuration?

Reply:
> I'm not sure this will work correctly if we update a list of 2 elements to a list of 2 different elements.

Actually I tested this and it works fine. If this was the case we would already have a problem with the current implementation because when you change a target configuration (let's say `message` for `targets.SnsTopic`) you are basically updating the target but keeping the same id. Rule targets are not CF resources by themselves, they are properties of `AWS::Events::Rule`.

You can try the following:
First depoy:
```ts
event.addTarget(new targets.SqsQueue(queue));
event.addTarget(new targets.SnsTopic(topic));
```
Second deploy:
```ts
event.addTarget(new targets.SnsTopic(topic));
event.addTarget(new targets.SqsQueue(queue));
```
After second deploy targets are correctly inverted.

Let me know if you still think that the id generation should be updated.

Reply:
Why do they have an ID then... :x